op-challenger/challenger/traversal.go

```diff
@@ -0,0 +1,247 @@
+		var err error
+		_, receipts, err = l.client.FetchReceipts(ctx, hash)
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+	return receipts, nil
+}
+
+// subscribeNewHead subscribes to new heads with a backoff.
+func (l *logTraversal) subscribeNewHead(ctx context.Context, headers chan *types.Header) (ethereum.Subscription, error) {
+	var sub ethereum.Subscription
+	err := backoff.DoCtx(ctx, 4, ExponentialBackoff, func() error {
+		ctx, cancel := context.WithTimeout(ctx, time.Second*5)
+		defer cancel()
+		var err error
+		sub, err = l.client.SubscribeNewHead(ctx, headers)
+		return err
+	})
+	return sub, err
+}
+
+// Quit quits the log traversal.
+func (l *logTraversal) Quit() {
+	l.quit <- struct{}{}
+	l.started = false
+}
+
+// Start starts the log traversal.
+func (l *logTraversal) Start(ctx context.Context, handleLog func(*types.Log) error) error {
+	if l.started {
+		return ErrAlreadyStarted
+	}
+	headers := make(chan *types.Header)
+	sub, err := l.subscribeNewHead(ctx, headers)
+	if err != nil {
+		l.log.Error("Failed to subscribe to new heads", "err", err)
+		return err
+	}
+	l.started = true
+	go l.onNewHead(ctx, headers, sub, handleLog)
+	return nil
+}
+
+// Started returns true if the log traversal has started.
+func (l *logTraversal) Started() bool {
+	return l.started
+}
+
+// onNewHead handles a new [types.Header].
+func (l *logTraversal) onNewHead(ctx context.Context, headers chan *types.Header, sub ethereum.Subscription, handleLog func(*types.Log) error) {
+	for {
+		select {
+		case <-l.quit:
+			l.log.Info("Stopping log traversal: received quit signal")
+			sub.Unsubscribe()
+			return
+		case header := <-headers:
+			l.log.Info("Received new head", "number", header.Number)
+			l.dispatchNewHead(ctx, header, handleLog, true)
+		}
+	}
+}
+
+// spawnCatchup spawns a new goroutine to "catchup" for missed/skipped blocks.
+func (l *logTraversal) spawnCatchup(ctx context.Context, start *big.Int, end *big.Int, handleLog func(*types.Log) error) {
+	for {
+		// Break if we've caught up (start > end).
+		if start.Cmp(end) == 1 {
+			l.log.Info("Traversal caught up", "start", start, "end", end)
+			return
+		}
+
+		info, err := l.fetchBlockInfoByNumber(ctx, start.Uint64())
+		if err != nil {
+			l.log.Error("Failed to fetch block", "err", err)
+			return
+		}
+
+		receipts, err := l.fetchBlockReceipts(ctx, info.Hash())
+		if err != nil {
+			l.log.Error("Failed to fetch receipts", "err", err)
+			return
+		}
+		for _, receipt := range receipts {
+			for _, log := range receipt.Logs {
+				err := handleLog(log)
+				if err != nil {
+					l.log.Error("Failed to handle log", "err", err)
+					return
+				}
+			}
+		}
+
+		// Increment to the next block to catch up to
+		start = start.Add(start, big.NewInt(1))
+	}
+}
+
+// updateBlockNumber updates the last block number with a mutex lock.
+func (l *logTraversal) updateBlockNumber(blockNumber *big.Int) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	l.lastBlockNumber = blockNumber
+}
```

Comment:
The locking around `lastBlockNumber` is inconsistent.  It's accessed without locks in `dispatchNewHead` and `LatestBlockNumber` but has locks around it when setting here. Both reads and writes need to use the same lock to ensure memory safety across threads.

But in this case, apart from the `LastBlockNumber` function which doesn't appear to be necessary, `lastBlockNumber` is only ever used from a single thread - the goroutine running `onNewHead`, so if `LastBlockNumber` were removed and the last block number treated entirely as an internal detail then no locks would be required for it.  Previously I think it was also being used by the `spawnCatchup` thread but that's not the case now.